server/log_operation_manager.go

```diff
@@ -90,124 +90,30 @@ type LogOperationInfo struct {
 	// batch takes longer than this interval to complete, the next batch
 	// will start immediately.
 	RunInterval time.Duration
-	// PreElectionPause is the maximum interval to wait before starting a
-	// mastership election for a particular log.
-	PreElectionPause time.Duration
-	// MasterCheckInterval is the interval between checks that we still
-	// hold mastership for a log.
-	MasterCheckInterval time.Duration
+
+	// ElectionConfig configures per-log master election process.
+	ElectionConfig election.Config
 	// MasterHoldInterval is the minimum interval to hold mastership for.
 	MasterHoldInterval time.Duration
-	// ResignOdds gives the chance of resigning mastership after each
-	// check interval, as the N for 1-in-N.
-	ResignOdds int
+	// ResignSpread is the max extra number of master check intervals to keep
+	// mastership for. The master will resign in:
+	//   MasterHoldInterval + rand(ResignSpread) * MasterCheckInverval.
+	ResignSpread int
+
 	// NumWorkers is the number of worker goroutines to run in parallel.
 	NumWorkers int
 }
 
-type electionRunner struct {
-	logID    int64
-	info     *LogOperationInfo
-	tracker  *election.MasterTracker
-	cancel   context.CancelFunc
-	wg       *sync.WaitGroup
-	election election.MasterElection
-}
-
-type resignation struct {
-	er   *electionRunner
-	done chan<- bool
-}
-
-func (r *resignation) execute(ctx context.Context) {
-	glog.Infof("%d: deliberately resigning mastership", r.er.logID)
-	if err := r.er.election.Resign(ctx); err != nil {
-		glog.Errorf("%d: failed to resign mastership: %v", r.er.logID, err)
+// getResignDelay computes master resignation delay based on operation
+// parameters.
+func getResignDelay(info *LogOperationInfo) time.Duration {
```

Comment:
This is not exactly equivalent to the previous `ResignOdds` behavior, although could be simpler?

The old one was a geometric distribution: the probability of resigning on `k`-th master check iteration was `(1-p)^(k-1) * p`, where `p=1/ResignOdds`. Expected number of iterations was `ResignOdds`. So the code here could look like:
```
	p := 1.0 / float64(info.ResignOdds)
	x := rand.Float64()
	checks := int64(math.Log2(1.0-x) / math.Log2(1.0-p))
	return delay + time.Duration(checks)*info.ElectionConfig.MasterCheckInterval
```

The new one is just uniform with expected number of extra waiting iterations equal to `ResignSpread/2` (could make it `ResignSpread` to be even more compatible with the old behaviour).

Reply:
I don't think the exact behaviour is important at this level. We want to be able to transfer mastership to allow things to balance out across multiple tasks but not too often and (probabilistically likely) not for all the logs at the same time.